pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,139 @@
+
+## The Trigger
+During the discussion, Sergio brought up that Fran was actually using an illegal reference, as he, in one of the examples, using a Reference Object for a server, which was not allowed.  More specifically, it was this example where he references the `mosquitto` server:
+
+```yaml
+asyncapi: 3.0.0
+...
+servers:
+  mosquitto:
+    $ref: 'common.asyncapi.yaml#/components/servers/mosquitto'
+```
+
+Immediately I said, "wait .... It's not?!". Cause before this, I had always used `$ref` quite extensively in my AsyncAPI documents and specifically used a reference for servers. And I knew that the tooling had no problems with the `$ref` as long as it was a valid reference. 
+
+But Sergio was absolutely right, a second look into the specification, `servers` are defined using [Servers Object](https://www.asyncapi.com/docs/specifications/v2.2.0#serversObject), which are defined using a map of [Server Object](https://www.asyncapi.com/docs/specifications/v2.2.0#serverObject)s. **NOT** `Server Object | Reference Object` as I expected.
+
+So after that, we started to realize, that there is quite a big difference when and where Reference Objects are allowed. For the full list of discrepancies, check out [spec #650](https://github.com/asyncapi/spec/issues/650).
+
+A quick side note, when I say tooling allowed it, I mean the [JS Parser](https://github.com/asyncapi/parser-js), as most of our tooling directly depend upon this library to parse and interact with AsyncAPI documents. 
+
+
+### But why did tooling allow it?!
+
+So back to my own experience, why was I so sure that the tooling allowed for me to use Reference Objects for servers? 
+
+Well, as it turns out, it is because that the underlying parser, bundles references before it validates the AsyncAPI document. This means that if I defined my AsyncAPI document such as:
+
+```yaml
+asyncapi: '2.2.0'
+info:
+  title: Test overriding dereferenced objects 
+  version: '1.0.0'
+servers:
+  test-server:
+    $ref: './servers/testServer.yaml'
+...
+```
+
+Together with `testServer.yaml`:
+```yaml
+url: ws://mycompany.com/ws
+protocol: ws
+```
+
+Validating the AsyncAPI document using the [JSON Schema representation for 2.2.0](https://github.com/asyncapi/spec-json-schemas/blob/master/schemas/2.2.0.json), would reject it. 
+
+However, because the parser bundles the reference together before doing so:
+
+```yaml
+asyncapi: '2.2.0'
+info:
+  title: Test overriding dereferenced objects 
+  version: '1.0.0'
+servers:
+  test-server:
+    url: ws://mycompany.com/ws
+    protocol: ws
+...
+```
+
+It is a fully valid AsyncAPI document once the parser tries to validate the document. Checkout [parser-js #405](https://github.com/asyncapi/parser-js/issues/405) for more information.
+
+## Deep diving
+We started discussing the reference behavior, and we started to peruse the specification in terms of the Reference Object and the [Schema Object](https://www.asyncapi.com/docs/specifications/v2.2.0#schemaObject).
```

Comment:
> peruse

I'm sure this is gonna happen again https://www.reddit.com/r/SiliconValleyHBO/comments/e1m6ik/google_trends_for_peruse/ :trollface: